Feat/breadcrumb

KrisKnabel opened the pull request:
The breadcrumb component is ready for review.  Design has taken a look at it and after feedback and refactoring, they have given it the thumbs up.  100% unit test coverage of the breadcrumb has been achieved with no accessibility errors.

KrisKnabel commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,129 @@
+      <span ref="ruler">
+        <span
+          v-for="(item, index) in items"
+          :key="index">
```
I was pretty sure the :key value scope is just for the collection but I could be wrong.  I would be surprised.  Better to be safe though so I like your mixin idea.

lancetharper replied:
I'm not sure if it's across the whole page, but using index alone is like not having a key at all since that's how vue would infer it anyway. At minimum I'd think it would need to have some contextual id like `breadcrumb${index}`

KrisKnabel replied:
So accroding to the docs and the vue.js forum, it is actually better and more performant to not provide a key in this instance.  If we had child view components, we want to define :key but since this is a simple v-for list, we should just use Vue.js default implementation.  I feel that we need to modify the eslint to allow for this.

KrisKnabel replied:
This has been modified to use the most performant Vue.js default key logic.

KrisKnabel commented on package.json:
```diff
@@ -9,6 +9,8 @@
     "commitmsg": "commitlint -e $GIT_PARAMS",
     "dev": "node build/dev-server.js",
+    "build": "cross-env NODE_ENV=prod node build/build.js",
+    "build:demo": "cross-env NODE_ENV=dev node build/build-demo.js",
```
It is for building the components to host for a quick demo or design review.  Not a true prod build, sort of a hybrid between prod and dev.  Now that I think of it, I should put this in a different branch.  I will go ahead and do that now.

lancetharper replied:
Ok, how is it different than `npm run dev`? I guess I'm not clear what a hybrid between dev and prod would do

KrisKnabel replied:
npm run dev does not produce a dist folder.  When you build for prod, you lose the source mapping and can't debug on the hosted site.  The build:demo builds the dist folder while still allowing the user to see readable files and can debug in the browser.

KrisKnabel replied:
I'm going to rip this out of this PR and keep it locally until it becomes useful again, say if design wants to look at changes on the fly or something.

KrisKnabel replied:
The demo build logic has been removed.

KrisKnabel commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,129 @@
+ */
+
+import modifier from 'mixinsdir/modifier';
+import _ from 'lodash';
```
Tried that and it was not working, I'll try again.

KrisKnabel replied:
Updated dependency to only grab debounce function from lodash.

mhewson commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,129 @@
+      <span
+        v-if="(truncate)"
+      >
+        <span class="breadcrumb-item ellipses"><a
```
here is a link the the start of our style guide - 
https://paper.dropbox.com/doc/REI-Cedar-code-style-guide-oHMNpqXcu1nvFLUI2pMUO

KrisKnabel replied:
Modified all class names to follow BEM guidelines

KrisKnabel commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,129 @@
+    </div>
+    <div class="breadcrumb-container">
+      <span
+        v-if="(truncate)"
```
Truncate is a computed property.  Made the change in response to your comments about not modifying the props directly.  Did I do it wrong?

lancetharper replied:
No, I just wondered if it needed the `()` around it. `truncate` vs `(truncate)`

KrisKnabel replied:
Ah, my guess is the parenthesis are not essential.  They were likely and remaining artifact from the previous comparison statement that was there before the refactor.  I will remove them.

KrisKnabel replied:
The parenthesis have been removed.

KrisKnabel commented on src/components/breadcrumb/__tests__/CdrBreadcrumb.spec.js:
```diff
@@ -0,0 +1,104 @@
+import debounce from 'lodash/debounce';
+
+//jest.unmock('lodash/debounce');
+debounce = jest.fn((fn) => fn);
```
This is breaking the unit test loader.  Still working on a way to get this to run.

KrisKnabel commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,129 @@
+        v-else
+        class="breadcrumb-item"
+        v-for="(item, index) in items"
+        :key="index">
```
Modified to use Vue.js default logic for key

KrisKnabel commented on src/components/breadcrumb/styles/CdrBreadcrumb.pcss:
```diff
@@ -0,0 +1,57 @@
+
+  /* Hover
+  ========== */
+  .breadcrumb-item {
```
Updated class names to follow BEM guidelines.

lancetharper commented:
I pushed a small change to flatten out the css a bit and fix that button background color.

Only thing left to do is refactor some of the markup to make it more a11y friendly (nav + ol) and add a couple aria tags.

Check [here](https://www.w3.org/TR/wai-aria-practices-1.1/examples/breadcrumb/index.html) for an example.

KrisKnabel commented:
Pushed changes suggested by Lance.  Ready for "hopefully" final review pass.

KrisKnabel commented:
Updated ellipses touch area, removed unnecessary div, updated unit tests, and updated example styling.  Please review.

KrisKnabel commented:
Updated ellipses to be a button with icon.  Only icon we had was vertically oriented so rotated it 90 degrees.  I was instructed not to translate it and keep it vertically centered(ish), but it looks odd IMHO.

lancetharper commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,142 @@
+  <nav
+    ref="container"
+    :class="modifierClass"
+    class="cdr-breadcrumb__container"
```
Doesn't look like this class actually does anything. Can probably just be removed since the base class is added here now too.

KrisKnabel replied:
Removed

lancetharper commented on src/components/breadcrumb/styles/CdrBreadcrumb.pcss:
```diff
@@ -0,0 +1,100 @@
+  /* Ellipses Icon
+  ========== */
+  &__ellipses-icon {
+    width: 1.4rem;
```
Overriding the icon size like this is load order dependent. If icon's css is loaded after breadcrumb, this won't do anything. We should be using the icons size modifier too rather than doing it this way.

KrisKnabel replied:
Made selector more specific.

KrisKnabel replied:
The icon size modifier was not working.  Setting the width and height directly was the only way I could figure out how to get the icon to show up at all.

lancetharper commented on src/components/breadcrumb/styles/CdrBreadcrumb.pcss:
```diff
@@ -0,0 +1,100 @@
+
+    /* Hover
+    ---------- */
+    &:hover .cdr-breadcrumb__link {
```
Can we get rid of this descendant selector since __link is set to inherit color anyway? __link needs to have `color: inherit` added for it's hover state to do this.

lancetharper replied:
active and focus states should almost always match hover too

KrisKnabel replied:
Removed descendant selector and updated active and focus states

lancetharper commented on src/components/breadcrumb/styles/CdrBreadcrumb.pcss:
```diff
@@ -0,0 +1,100 @@
+
+  /* Ellipses
+  ========== */
+  &__ellipses {
```
Looks like this should underline for hover/active/focus

KrisKnabel replied:
Had to add border-right since the svg does not respond to text-decoration and the svg was being rotated.

lancetharper commented:
Looks like the ellipse button needs screen reader text too to pass the a11y tests

KrisKnabel commented:
On Friday 6/15, pushed up SEO Refactor changes which included: 
  - Added ellipsis svg to use instead of rotated moreFunctions icon.
  - Hidden list items during truncation are now rendered and hidden so bots can still see them.
  - Modified breadcrumb collection to match JSON-LD format for breadcrumb for easier consumption
  - Updated examples to make it clearer to not use the current page/steps in breadcrumbs

lancetharper commented on src/components/breadcrumb/CdrBreadcrumb.vue:
```diff
@@ -0,0 +1,167 @@
+     *   url: location breadcrumb should navigate on click
+     * }
+     */
+    items: {
```
Can we get some validation for these props? `items` will probably need to check that the Array is of objects and that the objects have the required properties.

lancetharper replied:
I think we're all good after this 👍

KrisKnabel replied:
Validator function added for `items` prop.  The other props shouldn't need a validator function as the `required` and `type` properties should be sufficient.

KrisKnabel commented:
Done refactoring this breadcrumb component.  Ready for final review.

lancetharper commented:
Looks like this needs a new unit test still

KrisKnabel commented:
Added unit test for prop validator.  Now at 100%  coverage.

KrisKnabel commented:
Merging... Finally!

KrisKnabel commented:
Accidentally clicked close instead of merge button.
